src/common/type/pack.c

```diff
@@ -0,0 +1,1571 @@
+static size_t
+pckReadBuffer(PackRead *this, size_t size)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+        FUNCTION_TEST_PARAM(SIZE, size);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    size_t remaining = this->bufferUsed - this->bufferPos;
+
+    if (remaining < size)
+    {
+        if (this->read != NULL)
+        {
+            // Nothing can be remaining since each read fetches exactly the number of bytes required
+            ASSERT(remaining == 0);
+            bufUsedZero(this->buffer);
+
+            // Limit the buffer for the next read so we don't read past the end of the pack
+            bufLimitSet(this->buffer, size < bufSizeAlloc(this->buffer) ? size : bufSizeAlloc(this->buffer));
+
+            // Read bytes
+            ioReadSmall(this->read, this->buffer);
+            this->bufferPos = 0;
+            this->bufferUsed = bufUsed(this->buffer);
+            remaining = this->bufferUsed;
+        }
+
+        if (remaining < 1)
+            THROW(FormatError, "unexpected EOF");
+
+        FUNCTION_TEST_RETURN(remaining < size ? remaining : size);
+    }
+
+    FUNCTION_TEST_RETURN(size);
+}
+
+/***********************************************************************************************************************************
+Unpack an unsigned 64-bit integer from base-128 varint encoding
+***********************************************************************************************************************************/
+static uint64_t
+pckReadUInt64Internal(PackRead *this)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    uint64_t result = 0;
+    uint8_t byte;
+
+    for (unsigned int bufferIdx = 0; bufferIdx < PACK_UINT64_SIZE_MAX; bufferIdx++)
+    {
+        pckReadBuffer(this, 1);
+        byte = this->bufferPtr[this->bufferPos];
+
+        result |= (uint64_t)(byte & 0x7f) << (7 * bufferIdx);
+
+        if (byte < 0x80)
+            break;
+
+        this->bufferPos++;
+    }
+
+    if (byte >= 0x80)
+        THROW(FormatError, "unterminated base-128 integer");
+
+    this->bufferPos++;
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+/***********************************************************************************************************************************
+Read next field tag
+***********************************************************************************************************************************/
+static bool
+pckReadTagNext(PackRead *this)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_READ, this);
+    FUNCTION_TEST_END();
+
+    bool result = false;
+
+    // Read the tag byte
+    pckReadBuffer(this, 1);
+    unsigned int tag = this->bufferPtr[this->bufferPos];
+    this->bufferPos++;
+
+    // If the current container is complete (e.g. object)
+    if (tag == 0)
+    {
+        this->tagNextId = 0xFFFFFFFF;
```

Comment:
0xFFFFFFFF is used 4 times. I'd like to see a constant for it since it would be very easy to type 0xFFFFFFF accidentally.

Reply:
Agreed, and constant here would be good.

Reply:
Fixed in fa5bff95a5d51.